src/main/scala/cognite/spark/SdkV1Rdd.scala

```diff
@@ -24,16 +29,55 @@ case class SdkV1Rdd[A](
     new GenericClient[IO, Nothing](Constants.SparkDatasourceVersion)
 
   override def getPartitions: Array[Partition] = {
-    val numberOfStreams = getStreams(client, config.limit, numPartitions).length
-    0.until(numberOfStreams).toArray.map(CdfPartition)
+    val numberOfPartitions =
+      getStreams(client, config.limit, config.partitions, config.partitionsPerCompute).length
+    0.until(numberOfPartitions).toArray.map(CdfPartition)
   }
 
   override def compute(_split: Partition, context: TaskContext): Iterator[Row] = {
     val split = _split.asInstanceOf[CdfPartition]
+    val drainPool =
+      ExecutionContext.fromExecutor(Executors.newFixedThreadPool(1))
+    implicit val contextShift: ContextShift[IO] =
+      IO.contextShift(ExecutionContext.fromExecutor(Executors.newFixedThreadPool(1)))
+
+    val processedItems = mutable.Set[Long]()
+
+    val streams = getStreams(client, config.limit, config.partitions, config.partitionsPerCompute)
+      .drop(split.index)
+      .next
+      .reduce(_.merge(_))
+    val queue = new ArrayBlockingQueue[Vector[A]](20)
+
+    val pushIterator = for {
+      s <- streams.chunks
+      put <- {
+        val freshItems = s.toVector.filterNot(i => processedItems.contains(uniqueId(i)))
+        processedItems ++= freshItems.map(uniqueId)
+        Stream.eval(IO(queue.put(freshItems)))
+      }
+    } yield put
+
+    val f: Future[Unit] = Future {
+      pushIterator.compile.drain.unsafeRunSync()
+    }(drainPool)
+    new Iterator[Row] {
```

Comment:
I think we should move this logic to a separate helper class or function. It's quite hard to follow what's going on. It deserves a few comments, imo.

Reply:
Much better, thanks.